source/material_model/compositing.cc

```diff
@@ -0,0 +1,207 @@
+  along with ASPECT; see the file doc/COPYING.  If not see
+  <http://www.gnu.org/licenses/>.
+*/
+
+#include <deal.II/base/std_cxx11/array.h>
+#include <aspect/material_model/compositing.h>
+#include <utility>
+#include <limits>
+
+using namespace dealii;
+
+namespace aspect
+{
+  namespace MaterialModel
+  {
+    // Parse property names
+    template <int dim>
+    Property::MaterialProperty
+    Compositing<dim>::parse_property_name(const std::string &s)
+    {
+      if (Property::property_map.count(s) == 0)
+        AssertThrow (false,
+                     ExcMessage ("The value <" + s + "> for a material "
+                                 "property is not one of the valid values."));
+      return Property::viscosity;
+      return Property::property_map.at(s);
+    }
+
+    //Copy the requested data for one model
+    template <int dim>
+    void
+    Compositing<dim>::copy_required_properties(const unsigned int model_index,
+                                               const typename Interface<dim>::MaterialModelOutputs &base_output,
+                                               typename Interface<dim>::MaterialModelOutputs &out) const
+    {
+      if (model_property_map.at(Property::viscosity) == model_index)
+        out.viscosities = base_output.viscosities;
+      if (model_property_map.at(Property::density) == model_index)
+        out.densities = base_output.densities;
+      if (model_property_map.at(Property::thermal_expansion_coefficient) == model_index)
+        out.thermal_expansion_coefficients = base_output.thermal_expansion_coefficients;
+      if (model_property_map.at(Property::specific_heat) == model_index)
+        out.specific_heat = base_output.specific_heat;
+      if (model_property_map.at(Property::thermal_conductivity) == model_index)
+        out.thermal_conductivities = base_output.thermal_conductivities;
+      if (model_property_map.at(Property::compressibility) == model_index)
+        out.compressibilities = base_output.compressibilities;
+      if (model_property_map.at(Property::entropy_derivative_pressure) == model_index)
+        out.entropy_derivative_pressure = base_output.entropy_derivative_pressure;
+      if (model_property_map.at(Property::entropy_derivative_temperature) == model_index)
+        out.entropy_derivative_temperature = base_output.entropy_derivative_temperature;
+      if (model_property_map.at(Property::reaction_terms) == model_index)
+        out.reaction_terms = base_output.reaction_terms;
+    }
+
+    template <int dim>
+    void
+    Compositing<dim>::evaluate(const typename Interface<dim>::MaterialModelInputs &in,
+                               typename Interface<dim>::MaterialModelOutputs &out) const
+    {
+      typename Interface<dim>::MaterialModelOutputs base_output(out.viscosities.size(),
+                                                                this->introspection().n_compositional_fields);
```

Comment:
I had primarily been considering the memory use, but I had also been considering  whether there was a way to avoid the allocations entirely (Not possible while maintaining a consistent interface as compared to the other material models). This focus was driven by my knowledge of the attempts to avoid creating new MaterialModelInput objects in the major computational loops. (noted in the use of the new constructor issue).